### DISCUSSION TITLE
Converting a PDF of notes to dark mode

### DISCUSSION THREAD

**Discussion by daspri** [ImageMagick/ImageMagick#2137] [Help] [answered]
Given a PDF of notes with a white background, black text, with red & blue (or other color) diagrams I would like to make the notes "dark mode" with a background of RGB 32,32,32.

Inverted (negate) text is too light (grey) when on this background so it must brightened, but I want the colors to remain as they are but look good (brighter?) on the 32,32,32 background. If someone were to reference "the red line on page 2" I need the line to still be red, not its inverse, but the existing red would be too dull on a dark background.

Example PDF https://www.dropbox.com/s/2cl6jhyntolvqlc/Notes.pdf?dl=0

This hack is what I came up with but it is very slow (and is specific to this pdf with jpgs of a certain size) Is there a better way? (faster and more generic?)
The images are resized to be landscape letter paper size in the final PDF (after some trial and error, "-density 180" seems to work)
```
#!/bin/zsh
# 1. zsh is used for loop replacement variables
# 2. uses modulatecolor2.sh
# http://www.fmwconcepts.com/imagemagick/modulatecolor2/index.php
mkdir -p mog
#extract original images
pdfimages -all $1 orig
#rotate 90, negate and brighten the white writing
mogrify -path mog -rotate 90 -negate -modulate 130 orig*.jpg
mv mog/orig* .
#shift the original reds and blues back to red and blue, but shifted so they
#are more visible/look better on a 32,32,32 dark background
for f in ori*.jpg; do modulatecolor2.sh -c c -t 50 -u d -h 345 -s 200 -b 106  $f TEMP${f:r}.jpg ;done
for f in TEMP*.jpg; do modulatecolor2.sh -c y -t 50 -u d -h 202 -s 120 $f X${f:r}.jpg ;done
# set the background to 32,32,32 and resize as last step to retain quality until the end
mogrify -path mog -fuzz 4%  -fill 'rgb(32,32,32)' -opaque black -resize 45% X*.jpg
#convert to pdf
convert -density 180 mog/* dark.pdf
rm orig*.jpg
rm TEMP*.jpg
rm X*.jpg
rm -rf mog
```

**Comment by fmw42** [comment, marked as answer]
Try

```
line 1: rasterize the pdf at 150 dpi
line 2: increase the brightness
line 3: convert to HSV and increase the saturation
line 4: convert back to sRGB
line 5: replace white with gray
line 6: save the results

```
```
convert -density 150 Notes.pdf \
-level 0x60% \
-colorspace HSV -channel 1 -level 0x50% +channel \
-set colorspace HSV -colorspace sRGB \
-fill "srgb(32,32,32)" -opaque white \
Notes_%0d.png
```

Here is the first page processed:

![x-00](https://user-images.githubusercontent.com/13458218/84534621-204e4780-ac9f-11ea-9f95-f7cd3aac18fa.png)

**Comment by daspri** [comment]
Here is page 1 output from the above script.  I'd like to retain the smoothness of the ink strokes (it looks much less jagged than your example, symbols in math formulas could get pixelated out) and has more vibrant colors?

![p1](https://user-images.githubusercontent.com/4674343/84535180-21cc3f80-aca0-11ea-806c-fda5e0313362.jpg)

**Reply by fmw42** [reply]
Try this. It should be sharper. But I am not getting the white bright enough. Feel free to experiment further with the arguments or modify as desired.

convert -density 600 Notes.pdf \
-level 0x50% \
-colorspace HSV -channel 1 -level 0x25% +channel \
-set colorspace HSV -colorspace sRGB \
-fuzz 5% -fill "srgb(32,32,32)" -opaque white \
-resize 25% \
Notes_%02d.png


First page:

![Notes_00](https://user-images.githubusercontent.com/13458218/84536305-65c04400-aca2-11ea-932b-17560473f7ea.png)

**Reply by fmw42** [reply]
Here is another brighter version.

You can make it sharper by increasing the density and then resizing by the same factor. The level 0x20% controls the saturation. Smaller values than 20 will make it more saturated. You can brighten by increasing the 50 on the -brightness-contrast.

```
convert -density 600 Notes.pdf \
-auto-level \
-brightness-contrast 50,0 \
-colorspace HSV -channel 1 -level 0x20% +channel \
-set colorspace HSV -colorspace sRGB \
-fuzz 5% -fill "srgb(32,32,32)" -opaque white \
-resize 25% \
Notes_%02d.png
```

First page:

![Notes_00](https://user-images.githubusercontent.com/13458218/84536965-bedca780-aca3-11ea-861b-4c1540f9d48b.png)

**Reply by daspri** [reply]
Thank you.  I'll try to match the bright white ink in the one I posted.

**Reply by daspri** [reply]
When I use "ImageMagick 7.0.10-14 Q16 x86_64 2020-05-24" the above command leaves the background color white unless I remove the "-brightness-contrast" parameter. 
If I try "ImageMagick 6.9.7-4 Q16 x86_64 20170114" the above command turns both the blue and red to pink. 

Which version are you using?

**Reply by fmw42** [reply]
I was using IM 6.9.11.18. But I can confirm your issue on IM 7.0.10.19.  Here is a fix for the white background issue. The PDF renderer that I am using uses DEVICE=pngalpha, which processes transparent PDF file and leaves an alpha in this file. In IM 7, that needs to be disable with -alpha disassociate (equivalent to -alpha off in IM 6). But it then shows and issue that seems to be associated with -brightness-contrast. I have reported the issue.

 ```
magick -density 600 Notes.pdf[0] \
-auto-level \
-brightness-contrast 50,0 \
-colorspace HSV -channel 1 -level 0x20% +channel \
-set colorspace HSV -colorspace sRGB \
-alpha disassociate -fuzz 5% -fill "srgb(32,32,32)" -opaque white \
-resize 25% \
Notes7_%02d.png


```

**Comment by fmw42** [comment]
It turns out that the color issue is due to the use of HDRI in IM 7. It appears the increased brightness is pushing some pixels out of range and that is exaggerated by the colorspace change to HSV. So the result from -brightness-contrast needs to be clamped in HDRI.  So this command seems to work

```
magick -density 600 Notes.pdf[0] \
-auto-level \
-brightness-contrast 50,0 -clamp \
-colorspace HSV -channel 1 -level 0x20% +channel \
-set colorspace HSV -colorspace sRGB \
-alpha disassociate -fuzz 5% -fill "srgb(32,32,32)" -opaque white \
-resize 25% \
Notes7c_%02d.png

```

Notice the addition of -clamp and -alpha disassociate to fix the red/blue color issue and the background color issue, respectively.

![Notes7c_00](https://user-images.githubusercontent.com/13458218/84581824-0e050400-ad9a-11ea-8b79-3f6d12da7048.png)

**Reply by daspri** [reply]
Thank you, this works in IM7.   The challenge is brightening the white text without concurrently washing out the blue and red.  The colored lines don't look quite as good as in the image I posted (which is my goal image), and if you zoom in on the white text it seems to be choppier as well.  I'll try some different values, but maybe it just has to be done in separate steps as was done in the script using modulatecolor2.sh twice.

**Comment by fmw42** [comment]
Here is an alternate method.

```
magick -density 600 Notes.pdf[0] \
-auto-level \
-define modulate:colorspace=HSV -modulate 250,140,100 -clamp \
-alpha disassociate -fuzz 5% -fill "srgb(32,32,32)" -opaque white \
-resize 25% \
Notes7c_%02d.png
```

![Notes7c_00](https://user-images.githubusercontent.com/13458218/84605424-f721e880-ae51-11ea-8765-4f1b352f9790.png)
